src/ucp/core/ucp_ep.c

```diff
@@ -2319,15 +2329,31 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
     ucp_lane_index_t lane;
     ucs_status_t status;
 
+    if (ep->flags & UCP_EP_FLAG_FAILED) {
+        return;
+    }
+
     ucs_for_each_bit(lane, check_lanes) {
         ucs_assert(lane < UCP_MAX_LANES);
-        /* coverity[overrun-local] */
-        status = uct_ep_check(ep->uct_eps[lane], 0, NULL);
-        if (status == UCS_OK) {
-            *lane_map &= ~UCS_BIT(lane);
-        } else if (status != UCS_ERR_NO_RESOURCE) {
-            ucs_warn("unexpected return status from uct_ep_check(ep=%p): %s",
-                     ep, ucs_status_string(status));
+        /* in case if remote ID is defined, wireup is completed and EP is
+         * in connect-to-iface mode then use UCP/AM keepalive */
+        if ((ep->flags & UCP_EP_FLAG_REMOTE_ID) && /* we have remote ID to check */
+            (ep->flags & UCP_EP_FLAG_REMOTE_CONNECTED) && /* wireup completed */
+            (ucp_ep_get_iface_attr(ep, lane)->cap.flags &
+             UCT_IFACE_FLAG_CONNECT_TO_IFACE)) { /* connect-to-iface */
```

Comment:
maybe better have two different maps to avoid these check for every lane?

Reply:
actually same UCT EP may be processed as ep_check & AM ping-pong depends from EP state (connected or not)